### ISSUE TITLE
Constructing a STAT table

### ISSUE THREAD

**Issue by jenskutilek** [fonttools/fonttools#952]
Are there any examples for building a complete STAT table? I’m trying to compile one from a manually written TTX source, but I can’t seem to get the attribute names right.

```xml
  <STAT>
    <Version value="0x00010001"/>
    <DesignAxisRecordSize value="3"/>
    <!-- DesignAxisCount=4 -->
    <DesignAxisRecord>
      <Axis index="0">
        <AxisTag value="wght"/>
        <AxisNameID value="256"/>
        <AxisOrdering value="0"/>
      </Axis>
      <Axis index="1">
        <AxisTag value="wdth"/>
        <AxisNameID value="257"/>
        <AxisOrdering value="1"/>
      </Axis>
      <Axis index="1">
        <AxisTag value="opsz"/>
        <AxisNameID value="258"/>
        <AxisOrdering value="1"/>
      </Axis>
      <Axis index="2">
        <AxisTag value="ital"/>
        <AxisNameID value="259"/>
        <AxisOrdering value="2"/>
      </Axis>
    </DesignAxisRecord>
    <!-- AxisValueCount=1 -->
    <AxisValueArray>
      <!-- Regular -->
      <AxisValue>
        <Format value="3"/>
        <AxisIndex value="0"/>
        <Flags value="ElidableAxisValueName"/>
        <ValueNameID value="260"/>
        <Value value="400"/>
        <LinkedValue value="700"/>
      </AxisValue>
      <!-- Bold -->
      <AxisValue>
        <Format value="1"/>
        <AxisIndex value="0"/>
        <Flags/>
        <ValueNameID value="264"/>
        <Value value="700"/>
      </AxisValue>
      <!-- Italic -->
      <AxisValue>
        <Format value="3"/>
        <AxisIndex value="2"/>
        <Flags value="ElidableAxisValueName"/>
        <ValueNameID value="260"/>
        <Value value="0"/>
        <LinkedValue value="1"/>
      </AxisValue>
    </AxisValueArray>
    <ElidedFallbackNameID value="260"/>
  </STAT>
```

I get a traceback:

```
Parsing 'STAT' table...
ERROR: Unhandled exception has occurred
Traceback (most recent call last):
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttx.py", line 387, in main
    process(jobs, options)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttx.py", line 361, in process
    action(input, output, options)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/misc/loggingTools.py", line 372, in wrapper
    return func(*args, **kwds)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttx.py", line 272, in ttCompile
    ttf.importXML(input)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttLib/__init__.py", line 372, in importXML
    reader.read()
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/misc/xmlReader.py", line 48, in read
    self._parseFile(self.file)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/misc/xmlReader.py", line 73, in _parseFile
    parser.Parse(chunk, 0)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/misc/xmlReader.py", line 141, in _endElementHandler
    self.currentTable.fromXML(name, attrs, content, self.ttFont)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttLib/tables/otBase.py", line 100, in fromXML
    self.table.fromXML(name, attrs, content, font)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttLib/tables/otBase.py", line 738, in fromXML
    value = conv.xmlRead(attrs, content, font)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttLib/tables/otConverters.py", line 391, in xmlRead
    table.fromXML(name, attrs, content, font)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttLib/tables/otBase.py", line 738, in fromXML
    value = conv.xmlRead(attrs, content, font)
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttLib/tables/otConverters.py", line 378, in xmlRead
    for conv in table.getConverters():
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttLib/tables/otBase.py", line 772, in getConverters
    return self.converters[self.Format]
  File "/Users/jens/Library/Python/2.7/lib/python/site-packages/fontTools/ttLib/tables/otBase.py", line 539, in __getattr__
    raise AttributeError(attr)
AttributeError: Format
```

**Comment by robmck-ms** [comment]
Selawik has a STAT table that you can dump for an example. It's only one axis, though. (https://github.com/unicode-org/text-rendering-tests/blob/master/fonts/Selawik-variable.ttf).

Though not related to your error, I notice that you've got two Axis records with index="1" (wght, opsz).

Side question for @behdad and @brawer: Why do Axis records require an index parameter? Most other places in fontTools, records do not include their indices because fontTools computes it for us (and references are by name). It would be a lot nicer (and perhaps less prone to error for hand-written XML) if Axis records did not include an index, and AxisValue record's AxisIndex property would use the axis tag name instead of an index. (and then, perhaps, the property should be called Axis or AxisTag instead of AxisIndex).

**Comment by robmck-ms** [comment]
Here's a STAT table from a test font I'm working on. Note that I use numerics for Flags. I don't think there's support for enums like "ElidableAxisValueName", though that would be handy.

```
  <STAT>
    <Version value="0x00010001"/>
    <DesignAxisRecordSize value="8"/>
    <ElidedFallbackNameID value="2"/> <!-- "Small" -->
    <!-- DesignAxisCount=2 -->
    <DesignAxisRecord>
      <Axis index="0">
        <AxisTag value="ital"/>
        <AxisNameID value="256"/>
        <AxisOrdering value="2"/>
      </Axis>
      <Axis index="1">
        <AxisTag value="wght"/>
        <AxisNameID value="256"/>
        <AxisOrdering value="0"/>
      </Axis>
      <Axis index="2">
        <AxisTag value="opsz"/>
        <AxisNameID value="257"/>
        <AxisOrdering value="1"/>
      </Axis>
    </DesignAxisRecord>
    <!-- AxisValueCount=6 -->
    <AxisValueArray>
      <AxisValue index="0" Format="1">
        <AxisIndex value="0"/>
        <Flags value="2"/>
        <ValueNameID value="2"/>  <!-- Regular -->
        <Value value="0.0"/>
      </AxisValue>
      <AxisValue index="1" Format="1">
        <AxisIndex value="1"/>
        <Flags value="2"/>
        <ValueNameID value="2"/>  <!-- Regular -->
        <Value value="400.0"/>
      </AxisValue>
      <AxisValue index="2" Format="1">
        <AxisIndex value="1"/>
        <Flags value="0"/>
        <ValueNameID value="300"/>  <!-- Semibold -->
        <Value value="550.0"/>
      </AxisValue>
      <AxisValue index="3" Format="1">
        <AxisIndex value="1"/>
        <Flags value="0"/>
        <ValueNameID value="301"/>  <!-- Bold -->
        <Value value="700.0"/>
      </AxisValue>
      <AxisValue index="4" Format="2">
        <AxisIndex value="2"/>
        <Flags value="0"/>
        <ValueNameID value="258"/>  <!-- Small -->
        <NominalValue value="8.0"/>
        <RangeMinValue value="0.0"/>
        <RangeMaxValue value="9.7"/>
      </AxisValue>
      <AxisValue index="5" Format="2">
        <AxisIndex value="2"/>
        <Flags value="0"/>
        <ValueNameID value="259"/>  <!-- Text -->
        <NominalValue value="12.0"/>
        <RangeMinValue value="9.7"/>
        <RangeMaxValue value="13.5"/>
      </AxisValue>
      <AxisValue index="6" Format="2">
        <AxisIndex value="2"/>
        <Flags value="0"/>
        <ValueNameID value="260"/>  <!-- Subheading -->
        <NominalValue value="16.0"/>
        <RangeMinValue value="13.5"/>
        <RangeMaxValue value="18.5"/>
      </AxisValue>
      <AxisValue index="7" Format="2">
        <AxisIndex value="2"/>
        <Flags value="0"/>
        <ValueNameID value="261"/>  <!-- Heading -->
        <NominalValue value="20.0"/>
        <RangeMinValue value="18.5"/>
        <RangeMaxValue value="23.5"/>
      </AxisValue>
      <AxisValue index="8" Format="2">
        <AxisIndex value="2"/>
        <Flags value="0"/>
        <ValueNameID value="262"/>  <!-- Display -->
        <NominalValue value="26.0"/>
        <RangeMinValue value="23.5"/>
        <RangeMaxValue value="27.5"/>
      </AxisValue>
      <AxisValue index="9" Format="2">
        <AxisIndex value="2"/>
        <Flags value="0"/>
        <ValueNameID value="263"/>  <!-- Banner -->
        <NominalValue value="36.0"/>
        <RangeMinValue value="27.5"/>
        <RangeMaxValue value="32767.99998"/> <!-- represents infinity -->
      </AxisValue>
    </AxisValueArray>
  </STAT>

```

**Comment by behdad** [comment]
> Side question for @behdad and @brawer: Why do Axis records require an index parameter?

The index is printed FYI only.  FontTools does that whenever an array is converted to XML.

> Most other places in fontTools, records do not include their indices because fontTools computes it for us (and references are by name). It would be a lot nicer (and perhaps less prone to error for hand-written XML) if Axis records did not include an index, and AxisValue record's AxisIndex property would use the axis tag name instead of an index. (and then, perhaps, the property should be called Axis or AxisTag instead of AxisIndex).

That would be convenient, but would require handcoding the table, which I'd rather we do not do...  This table is hard to follow already....

**Comment by jenskutilek** [comment]
Thank you @robmck-ms, looking at the Selawik STAT table was helpful.

**Comment by anthrotype** [comment]
let's keep this open until this is done:
https://github.com/LettError/designSpaceDocument/issues/8

**Comment by davelab6** [comment]
Seems related to https://github.com/fonttools/fonttools/issues/1072

**Comment by anthrotype** [comment]
The OpenType Feature File spec has been updated to support [defining STAT table](http://adobe-type-tools.github.io/afdko/OpenTypeFeatureFileSpecification.html#9i-stat-table), and the latest makeotf from afdko 3.4.0 added support for building that (https://github.com/adobe-type-tools/afdko/pull/1127).
feaLib should be updated to also support this. Any takers?

**Comment by davelab6** [comment]
@simoncozens ?

**Comment by simoncozens** [comment]
Maybe me. Khaled did the makeotf implementation, but I'm happy to take this if he doesn't want it.

**Comment by khaledhosny** [comment]
I might give it a try if no one beat me to it, but right now I’m a bit busy with other stuff.

**Comment by punchcutter** [comment]
I hope nobody did anything yet because I'm basically done with this. Just need to clean up a bit and add more tests, but so far I got it working the same as `makeotf`.

**Comment by jenskutilek** [comment]
I think this is done now.
